src/main/java/uk/gov/nationalarchives/pdi/step/jena/groupmerge/JenaGroupMergeStep.java

```diff
@@ -56,226 +60,337 @@ public boolean processRow(final StepMetaInterface smi, final StepDataInterface s
-                    final LinkedHashMap<String, Model> currentGroupModels = new LinkedHashMap<>(currentRowModels.size());
-
-                    // should we mutate the first model in the group of rows, or create a new model and copy it in?
-                    for (final FieldModel currentRowModel : currentRowModels) {
-                        final Model model;
-                        if(currentRowModel.model == null) {
-                            // may be created later in mergeModels
-                            model = null;
-
-                        } else if (meta.isMutateFirstModel()) {
-                            // mutate the first model
-                            model = currentRowModel.model;
-
-                        } else {
-                            // create a new model
-                            model = ModelFactory.createDefaultModel();
-                            model.add(currentRowModel.model);
-
-                            /*
-                               Model can no longer be used if
-                               the user chose to remove it from the available
-                               output fields... so close it!
-                             */
-                            if (meta.isRemoveSelectedFields()) {
-                                currentRowModel.model.close();
-                            }
-                        }
-
-                        currentGroupModels.put(currentRowModel.fieldName, model);
-                    }
+    /**
+     * Process an input row as 1+n row in a group.
+     *
+     * @param meta this steps meta.
+     * @param data this steps data.
+     * @param inputRowData the input row.
+     */
+    static void mergeRowIntoGroup(final JenaGroupMergeStepMeta meta, final JenaGroupMergeStepData data, final Object[] inputRowData) {
+        // get the array holding the current Group Merged Row
+        final Object[] outputRowData = data.getGroupMergedRow();
+
+        // function - get the model of the current targetField in the Group Merged Row
+        final BiFunction<Object[], Integer, Model> fnGetOutputRowTargetFieldModel = (outputRowData1, outputRowTargetFieldIndex) -> (Model) outputRowData1[outputRowTargetFieldIndex];
+
+        // function - get the value of a normal field from the input row for the output row
+        final QuadFunction<OtherFieldAction, Object, Object[], Integer, Object> fnGetNormalFieldOutputValue = (otherFieldAction, inputRowFieldValue, outputRowData1, outputRowFieldIndex) -> {
+            if (otherFieldAction == OtherFieldAction.USE_LAST) {
+                return inputRowFieldValue;
+            } else if (otherFieldAction == OtherFieldAction.SET_NULL) {
+                return null;
+            } else if (otherFieldAction == OtherFieldAction.NULL_IF_DIFFERENT &&
+                    !outputRowData1[outputRowFieldIndex].equals(inputRowFieldValue)) {
```

Comment:
I believe this can result in a `NullPointerException` if the value at `outputRowData1[outputRowFieldIndex]` is `null`

Reply:
Ah ha! Good catch! I have pushed a commit now to fix that.